crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,235 @@
+            callback(result.map(|buffer| js_sys::Uint8Array::new(&buffer).to_vec()));
+        })
+    }
+
+    /// Generic function to start the async read of the `FileReader`.
+    ///
+    /// `callback` is the user-supplied callback, `extract_fn` function converts between JsValue
+    /// returned by the read and the type the callback expects, and `read_fn` is the method that
+    /// runs the async read on the `FileReader`.
+    fn read<T, CF, EF, RF>(blob: &Blob, callback: CF, extract_fn: EF, read_fn: RF) -> FileReader
+    where
+        CF: FnOnce(Result<T, FileReadError>) + 'static,
+        EF: Fn(JsValue) -> T + 'static,
+        RF: Fn(&web_sys::FileReader, &web_sys::Blob) + 'static,
+    {
+        // we need to be able to run the FnOnce, while proving to the compiler that it can only run
+        // once. The easiest way is to `take` it out of an Option. The `Rc` and `RefCell` are
+        // because we need shared ownership and mutability (for the `take`).
+        let load_callback: Rc<RefCell<Option<CF>>> = Rc::new(RefCell::new(Some(callback)));
+        let error_callback = load_callback.clone();
+
+        let reader = web_sys::FileReader::new().unwrap_throw();
+
+        let load_reader = reader.clone();
+        let error_reader = reader.clone();
+        // Either the load listener or the error listener will be called, never both (so FnOnce is
+        // ok).
+        let load_listener = EventListener::new(&reader, "load", move |_event| {
+            let result = extract_fn(load_reader.result().unwrap_throw());
+            let callback = load_callback.borrow_mut().take().unwrap_throw();
+            callback(Ok(result));
+        });
+
+        let error_listener = EventListener::new(&reader, "error", move |_event| {
+            let exception = error_reader.error().unwrap_throw();
+            let error = match exception.name().as_str() {
+                "NotFoundError" => FileReadError::NotFound(exception.message()),
+                "NotReadableError" => FileReadError::NotReadable(exception.message()),
+                "SecurityError" => FileReadError::Security(exception.message()),
+                // This branch should never be hit, so returning a less helpful error message is
+                // less of an issue than pulling in `format!` code.
+                _ => throw_str("unrecognised error type"),
+            };
+            let callback = error_callback.borrow_mut().take().unwrap_throw();
+            callback(Err(error));
+        });
+
+        read_fn(&reader, blob.as_ref());
+
+        FileReader {
+            reader,
+            load_listener,
+            error_listener,
+        }
+    }
+}
+
+#[cfg(feature = "futures")]
+pub mod futures {
+    use crate::{Blob, FileReadError};
+    use wasm_bindgen::UnwrapThrowExt;
+
+    /// Returns the contents of `blob` as a text string.
+    pub async fn read_as_text(blob: &Blob) -> Result<String, FileReadError> {
```

Comment:
By the way, I just realized that these `async` functions capture the lifetime for their argument, which isn't good. So they should be implemented like this instead:

```rust
pub fn read_as_text(blob: &Blob) -> impl Future<Output = Result<String, FileReadError>> {
    let (sender, receiver) = futures_channel::oneshot::channel();

    let reader = super::callbacks::read_as_text(blob, move |result| {
        sender.send(result).unwrap_throw();
    });

    async move {
        let output = receiver.await.unwrap_throw();
        drop(reader);
        output
    }
}
```